src/Extension/FrontMatter/FrontMatterParser.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the league/commonmark package.
+ *
+ * (c) Colin O'Dell <colinodell@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace League\CommonMark\Extension\FrontMatter;
+
+use League\CommonMark\Event\DocumentPreParsedEvent;
+use League\CommonMark\Extension\FrontMatter\Yaml\FrontMatterParserInterface;
+use League\CommonMark\Input\MarkdownInput;
+use League\CommonMark\Parser\Cursor;
+
+final class FrontMatterParser
+{
+    /**
+     * @var FrontMatterParserInterface
+     *
+     * @psalm-readonly
+     */
+    private $yamlParser;
+
+    private const REGEX_FRONT_MATTER = '/^---\\n.*\\n---\n/s';
+
+    public function __construct(FrontMatterParserInterface $yamlParser)
+    {
+        $this->yamlParser = $yamlParser;
+    }
+
+    public function __invoke(DocumentPreParsedEvent $event): void
+    {
+        $content = $event->getMarkdown()->getContent();
+
+        $cursor = new Cursor($content);
+
+        // Locate the front matter
+        $frontMatter = $cursor->match(self::REGEX_FRONT_MATTER);
+        if ($frontMatter === null) {
+            return;
+        }
+
+        // Trim the last 4 characters (ending ---s and newline)
+        $frontMatter = \substr($frontMatter, 0, -4);
+
+        // Parse the resulting YAML data
+        $data = $this->yamlParser->parse($frontMatter);
+
+        // Store the parsed data in the Document
+        $event->getDocument()->data['front_matter'] = $data;
```

Comment:
It would be really nice if this also captured the line offset of the front matter.

```php
// Determine the real source line by counting all newlines in the first
// match (which includes the front matter separators) and append a new
// line to denote that the content should start after it.
$line = preg_match_all('/\R/', $frontMatter . "\n") + 1;
```

Reply:
Implemented via ec7c4615